blob/conformance.go

```diff
@@ -0,0 +1,460 @@
+				if err != nil {
+					t.Errorf("unexpected error during read: %v", err)
+				}
+				if int64(n) != tc.wantReadSize {
+					t.Errorf("got read length %d want %d", n, tc.wantReadSize)
+				}
+				if !cmp.Equal(got[:tc.wantReadSize], tc.want) {
+					t.Errorf("got %q want %q", string(got), string(tc.want))
+				}
+				// TODO(issue #305): Fails for S3.
+				/*
+					if r.Size() != contentSize {
+						t.Errorf("got size %d want %d", r.Size(), contentSize)
+					}
+				*/
+				r.Close()
+			})
+		}
+	})
+}
+
+// testAttributes tests the behavior of attributes returned by Reader.
+func testAttributes(t *testing.T, makeBkt bucketMaker) {
+	const (
+		key         = "blob-for-attributes"
+		contentType = "text/plain"
+	)
+	content := []byte("Hello World!")
+
+	ctx := context.Background()
+	t.Run("Attributes", func(t *testing.T) {
+
+		// Create a blob to be read by sub-tests below.
+		b, done := makeBkt(t)
+		defer done()
+
+		opts := &WriterOptions{
+			ContentType: contentType,
+		}
+		w, err := b.NewWriter(ctx, key, opts)
+		if err != nil {
+			t.Fatal(err)
+		}
+		_, err = w.Write(content)
+		if err == nil {
+			err = w.Close()
+		}
+		if err != nil {
+			t.Fatal(err)
+		}
+		defer func() {
+			_ = b.Delete(ctx, key)
+		}()
+
+		t.Run("ContentType", func(t *testing.T) {
+			r, err := b.NewRangeReader(ctx, key, 0, 0)
+			if err != nil {
+				t.Fatalf("failed NewRangeReader: %v", err)
+			}
+			defer r.Close()
+			if r.ContentType() != contentType {
+				t.Errorf("got ContentType %q want %q", r.ContentType(), contentType)
+			}
+		})
+
+		// TODO(issue #303): Fails for GCS.
+		/*
+			t.Run("Size", func(t *testing.T) {
+				r, err := b.NewRangeReader(ctx, key, 0, 0)
+				if err != nil {
+					t.Errorf("failed NewRangeReader: %v", err)
+				}
+				defer r.Close()
+				if r.Size() != int64(len(content)) {
+					t.Errorf("got Size %d want %d", r.Size(), len(content))
+				}
+			})
+		*/
+
+		t.Run("ModTime", func(t *testing.T) {
+			r, err := b.NewRangeReader(ctx, key, 0, 0)
+			if err != nil {
+				t.Fatalf("failed NewRangeReader: %v", err)
+			}
+			defer r.Close()
+			t1 := r.ModTime()
+			if t1.IsZero() {
+				// This provider doesn't support ModTime.
+				return
+			}
+			// Touch the file after a couple of seconds and make sure ModTime changes.
+			time.Sleep(2 * time.Second)
+			w, err := b.NewWriter(ctx, key, nil)
+			if err != nil {
+				t.Fatalf("failed NewWriter: %v", err)
+			}
+			if err = w.Close(); err != nil {
+				t.Errorf("failed NewWriter Close: %v", err)
+			}
+			r2, err := b.NewRangeReader(ctx, key, 0, 0)
+			if err != nil {
+				t.Errorf("failed NewRangeReader#2: %v", err)
+			}
+			defer r2.Close()
+			t2 := r2.ModTime()
+			if !t2.After(t1) {
+				t.Errorf("ModTime %v is not after %v", t2, t1)
+			}
+		})
+	})
+}
+
+// loadTestFile loads a file from the blob/testdata/ directory.
+func loadTestFile(t *testing.T, filename string) []byte {
+	data, err := ioutil.ReadFile(filepath.Join("..", "testdata", filename))
```

Comment:
This would not work for running the conformance test on packages outside of our repository. That seems like a non-starter to me.

Reply:
Yea, you are right. OK to make this a TODO to fix? Any suggestions?